### PR TITLE
Upgrade to arrow 22

### LINKED ISSUE #3362: Upgrade to Arrow 22
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I would like DataFusion to use Arrow 22 so that I can proxy Flights in Ballista.

**Describe the solution you'd like**

Upgrade to arrow v22

**Describe alternatives you've considered**
N/A

### PR DESCRIPTION
**Pull request by avantgardnerio**
# Which issue does this PR close?

Closes #3362.

 # Rationale for this change

We should keep DataFusion up to date with Arrow.

# What changes are included in this PR?

- precision is always u8 now
- builders are in the middle of being refactored so some have `new()` and others `with_capacity()`
- pyo3 needs to be upgraded to 0.17

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by avantgardnerio** [comment]
@alamb thanks for releasing arrow 22! Hopefully this helps release the next version of DataFusion :)

File: datafusion-examples/examples/custom_datasource.rs [hunk 1]
```diff
@@ -243,8 +243,8 @@ impl ExecutionPlan for CustomExec {
             db.data.values().cloned().collect()
         };
 
-        let mut id_array = UInt8Builder::new(users.len());
-        let mut account_array = UInt64Builder::new(users.len());
+        let mut id_array = UInt8Builder::new();
```
Comment (yjshen) [inline datafusion-examples/examples/custom_datasource.rs:246]: `with_capacity` instead?

File: datafusion/common/Cargo.toml [hunk 2]
```diff
@@ -39,12 +39,12 @@ pyarrow = ["pyo3"]
...
 apache-avro = { version = "0.14", features = ["snappy"], optional = true }
-arrow = { version = "21.0.0", features = ["prettyprint"] }
+arrow = { version = "22.0.0", features = ["prettyprint"] }
 avro-rs = { version = "0.13", features = ["snappy"], optional = true }
 cranelift-module = { version = "0.87.0", optional = true }
 object_store = { version = "0.4.0", optional = true }
 ordered-float = "3.0"
-parquet = { version = "21.0.0", optional = true, features = ["arrow"] }
-pyo3 = { version = "0.16", optional = true }
+parquet = { features = ["arrow"], optional = true, version = "22.0.0" }
```
Comment (yjshen) [inline datafusion/common/Cargo.toml:47]: It seems to be a common practice to always have the version field at first

File: datafusion/core/src/physical_plan/explain.rs [hunk 3]
```diff
@@ -121,8 +121,8 @@ impl ExecutionPlan for ExplainExec {
             )));
         }
 
-        let mut type_builder = StringBuilder::new(self.stringified_plans.len());
-        let mut plan_builder = StringBuilder::new(self.stringified_plans.len());
+        let mut type_builder = StringBuilder::new();
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/explain.rs:124]: 👍 A lot of misuse of the previous version of the StringBuilder API: the capacity of StringBuilder is mistakenly taken as the number of items in the array.

How do you think if we use `StringBuilder::with_capacity(self.stringified_plans.len(), 1024)`, seems mirror the intentions clearer.

**Review by yjshen — COMMENTED** [review]
I'm slightly concerned about discarding the `StringBuilder` capacity. `StringBuilder::with_capacity(num_of_items, 1024)` seems a better choice for me.

File: datafusion-examples/examples/custom_datasource.rs (hunk 1 above)
Comment (avantgardnerio) [reply datafusion-examples/examples/custom_datasource.rs:246]: I'm happy to put these all back to `with_capacity()`, but now that they take two parameters, I wasn't sure what to set the `data_size` to. (`item_count` is straight-forward).

File: datafusion/common/Cargo.toml (hunk 2 above)
Comment (avantgardnerio) [reply datafusion/common/Cargo.toml:47]: This was done by an automated tool. I'll fix my tool and re-submit.

File: datafusion/core/src/physical_plan/explain.rs (hunk 3 above)
Comment (avantgardnerio) [reply datafusion/core/src/physical_plan/explain.rs:124]: So always `1024` when we don't know? Lack of knowledge of a good default was the only reason I didn't do that everywhere.

File: datafusion-examples/examples/custom_datasource.rs (hunk 1 above)
Comment (yjshen) [reply datafusion-examples/examples/custom_datasource.rs:246]: UInt8Builder is for primitive type UInt8, and its `with_capacity` takes only one parameter?
```rs
/// Creates a new primitive array builder with capacity no of items
    pub fn with_capacity(capacity: usize) -> Self {
```

File: datafusion/common/src/scalar.rs [hunk 4]
```diff
@@ -460,6 +460,7 @@ macro_rules! typed_cast {
     }};
 }
 
+// keep until https://github.com/apache/arrow-rs/issues/2054 is finished
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:463]: ❤️  for the comment

File: datafusion/core/src/avro_to_arrow/arrow_array_reader.rs [hunk 5]
```diff
@@ -258,7 +258,7 @@ impl<'a, R: Read> AvroArrowArrayReader<'a, R> {
     {
         let mut builder: Box<dyn ArrayBuilder> = match data_type {
             DataType::Utf8 => {
-                let values_builder = StringBuilder::new(rows.len() * 5);
+                let values_builder = StringBuilder::with_capacity(rows.len(), 5);
```
Comment (alamb) [inline datafusion/core/src/avro_to_arrow/arrow_array_reader.rs:261]: 👍

File: datafusion/core/src/physical_plan/common.rs [hunk 6]
```diff
@@ -393,7 +393,10 @@ mod tests {
             ]),
         };
 
-        assert_eq!(result, expected);
+        // Prevent test flakiness due to undefined / changing implementation details
```
Comment (alamb) [inline datafusion/core/src/physical_plan/common.rs:396]: This makes sense (and is something we see in IOx as well -- that the exact size of the parquet file changes from release to release). I think at least part of the discrepancy is that the version and gitsha of the arrow-rs library are encoded in the parquet metadata

File: datafusion/core/src/physical_plan/explain.rs (hunk 3 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/explain.rs:124]: FWIW I don't think the performance of explain plan is particularly important -- a few more allocations likely won't hurt either way

**Review by alamb — APPROVED** [review]
LGTM -- thank you @avantgardnerio ❤️ 

Let me know if you want to make any additional changes otherwise I will plan to merge this in later today or tomorrow

**Review by Dandandan — APPROVED** [review]

File: datafusion-examples/examples/custom_datasource.rs (hunk 1 above)
Comment (avantgardnerio) [reply datafusion-examples/examples/custom_datasource.rs]: You're right - I just got sloppy after fixing a few dozen of these. This is resolved now.

File: datafusion/core/src/physical_plan/explain.rs (hunk 3 above)
Comment (avantgardnerio) [reply datafusion/core/src/physical_plan/explain.rs]: I updated them all there it was sensible with a default of `1024`.

**Review by yjshen — APPROVED** [review]
LGTM, Thanks @avantgardnerio!

File: datafusion/common/Cargo.toml (hunk 2 above)
Comment (avantgardnerio) [reply datafusion/common/Cargo.toml]: Resolved with this ordering: https://github.com/avantgardnerio/workspace-gen/blob/891699e8cd4d85c7349cf70518a1f18628ea570d/src/main.rs#L242

**Comment by avantgardnerio** [comment]
CI seems broken. @alamb or @andygrove appreciate it if you could kick it please.

**Comment by avantgardnerio** [comment]
Okay, ready for merge. No idea why github is taking so long today...

**Comment by andygrove** [comment]
Thanks @avantgardnerio!

**Pull request merged**